Benchmarking RFC (#1063)
o1-labs/proof-systems: book/src/rfcs/benchmarking.md

# Benchmarking RFC

This document proposes a high level approach to benchmarking the crypto stack, mostly kimchi, snarky and pickles. The main concern is to set up an infrastructure where it is possible to integrate different kinds of benchmarks with minimal effort. While the specific benchmarks are not an immediate concern, a few benchmarks of interest will be mentioned later.

## Types of benchmarks

> There are different metrics that we can collect, and different ways of presenting them depending of who is interested in them.

Author: done

### Audience

We can differentiate to groups of people interested in benchmarks:

- Developers working in the crypto stack, mostly our crypto team, where the main interest is to avoid regressions or make performance improvements.
- People working outside of it, mostly users of our libraries, but could also include decision makers comparing other options or our marketing team, here the interest is more about more general metrics that can be compared with the competition or used to learn how to most efficiently use our technology.

Must be mentioned that some overlap should be expected, some benchmarks may be useful for anyone while other just to some.

### Metrics

> There are two main metrics, performance and memory, and a smaller one hardware utilization.

Reviewer: ```suggestion
Three metrics are considered: performance, memory, and hardware utilization. The main metrics are performance and memory. The hardware utilization metric has a more limited scope.
```

> There are two main metrics, performance and memory, and a smaller one hardware utilization.

Author: done

#### Performance

Basically execution time, this may be the hardest to measure, because the environment can have a considerable effect over the results, some high quality machine would be required, ideally a fully dedicated one to minimize noise, there is a way of measuring that overcomes noise problems that will be mentioned.
It can be subdivided it 2 other metrics or alternative approaches:

> - Execution time: The most simple, run the program in question several times and take the average, it may become quite expensive if the program measured takes some time because that time will be multiplied by the number of samples, in those cases it can be made faster sacrificing precision. This benchmark is one of the most interesting for external users, and one of the main comparison point for evaluating the competition. It can also serve for internal developer as a regression test or to verify performance improvements.

Reviewer: ```suggestion
: Execution time
The simplest approach. It runs the program several times and returns the overall average execution time. This benchmark might become slow in cases where the program takes too much time to run. In such scenarios, this benchmark can be made faster by reducing the number of samples (i.e., worst precision) it uses. This approach is one of the most interesting for external users and a key datapoint to compare how our proof-system fares against competitors. The execution time metric can also help our engineering teams as a metric to identify regressions or to verify performance improvements.
```

> - CPU instructions: This method instead of recording execution time counts the number of cpu instructions used, not very useful for external developer given that it doesn't translate directly to time. It is mostly useful to internal developer as an alternative to execution time, it is faster given that you only have to run it once to collect the information and can be run on cheap machines without significative noise given that execution time doesn't affect the number of instructions. Additionally the tools used to measure it also collect some other metrics that can be useful to aide optimization like cache misses and branch mispredictions.

Reviewer: ```suggestion
: CPU instructions
Instead of recording execution time, this method counts the number of cpu instructions used in a program execution, not very useful for external developer given that it doesn't translate directly to time. It is mostly useful to our engineering teams as an alternative to execution time, it is faster given that you only have to run it once to collect the information, and can be run on cheap machines without significative noise given that execution time doesn't affect the number of instructions. Additionally, the tools used to measure it also collect some other metrics that can be useful to aide optimization like cache misses and branch mispredictions.
```

It is easier to measure, as long as there is enough memory, noise shouldn't affect the results.

#### Hardware utilization

It can be of interest in some cases to measure how much of the available resources are actually used, ideally the same computer will take half the time with twice the resources, unfortunately this isn't always the case and sometimes additional resources go unused.
This doesn't change that often, the main utility of these benchmarks is just to detect low utilization so that it can be improved.
Hardware utilization could include many things, GPU acceleration for example if we were to start using it, but for now will only include CPU/core utilization.

### Layers

There are 3 main layers of the stack that we want to consider, kimchi, snarky and pickles. Given that they are implemented in different languages and exist in different repositories special considerations each layer will be required.

## Where to run the benchmarks

> To start everyone with the resources should be able to run the benchmarks locally, additionally we may want some of the benchmarks to run in CI to detect regressions. Some benchmarks may be considerably expensive to run, a decision will have to be made about which benchmarks to run and how often we should run them.

Reviewer 2: Detecting regression in performance might also include thinking more about CI infrastructure than other benchmarks. Both for very short or very long tests.

## Parameters

> We may be interested in parameterizing some benchmark, for example benchmarking verification time but doing it with different circuits sizes, that would show how performance scales with the size of the circuit.

Reviewer 2: ```suggestion
We may be interested in parameterizing some benchmarks, for example benchmarking verification time but doing it with different circuits sizes, that would show how performance scales with the size of the circuit.
```

Reviewer 2: I think this is a need, not a possible interest.

| kimchi time              | proof-systems | time     | yes              | yes              | yes             | yes              |
| kimchi time flamegraph   | proof-systems | time     | no               | yes              | no              | no               |
| kimchi instructions      | proof-systems | time     | no               | yes              | yes             | no               |
| kimchi memory            | proof-systems | memory   | yes              | yes              | yes             | no               |
| kimchi memory flamegraph | proof-systems | memory   | no               | yes              | no              | no               |
| snarky time              | mina/snarky*  | time     | yes              | yes              | yes             | yes              |
| snarky memory            | mina/snarky*  | memory   | yes              | yes              | yes             | no               |
| pickles time             | mina/pickles* | time     | yes              | yes              | yes             | yes              |
| pickles memory           | mina/pickles* | memory   | yes              | yes              | yes             | no               |

> \* these have to be in Mina now because pickles is in Mina, and while snarky has it's repo, to use it requires a few things still in Mina, if it were possible change that it would make these benchmarks easier and probably cheaper.

Reviewer 2: ```suggestion
\* these have to be in Mina now because pickles is in Mina, and while snarky has its own repository, using it requires a few things still in Mina, so that we might want to consider some changes to make these benchmarks easier and probably cheaper.
```

Reviewer 2: Could you exemplify 1 or 2 changes that would be required?

### Kimchi time

> One of the simplest measuring running time, can be an specific case or allow more flexibility by introducing parameters like for example different circuit sizes. We already have most of the infrastructure for this with criterion. Would ideally have a dedicated computer, considering that there aren't that many PRs in proof-systems the requirements shouldn't be that high.

Reviewer 2: ```suggestion
One of the simplest measuring running time, can be a specific case or allow more flexibility by introducing parameters like for example different circuit sizes. We already have most of the infrastructure for this with criterion. We would ideally have a dedicated computer. However, considering that there aren't that many PRs in proof-systems, the requirements shouldn't be that high.
```

Reviewer 2: Link to criterion?

Author: done

### Kimchi time flamegraph

> This can be considered a variant of the previous, some more research is required but ideally can be generated from the already existing benchmarks with little to no changes.

Reviewer 2: ```suggestion
This can be considered a variant of the previous. Some more research is required but this information ideally can be generated from the already existing benchmarks with little to no changes.
```

Reviewer 2: Link to previous section

> This may not be as interesting to have in CI, but can be useful to run locally for those working in optimizations.

Reviewer 2: I agree.

### Kimchi instructions

> This is basically counting the CPU instructions that takes to run some code, could be use as an alternative for CI given that it won't be affected by noise and can work in any computer. It would also be faster by running the code only once. The implementation will require a bit more research but one option is the crate iai, with some thinking it could be possible to have this and the time benchmarks share the same code.

Reviewer 2: Link to `iai` https://crates.io/crates/iai ?. Counting instructions is nice but counting CPU cycles might make more sense, especially on a CISCI-like architecture like `x86. However I see that iai can do that. Should we switch the instructions target to the cycles one?  Or be more generic wrt to the info that `iai` can generate.

Is worth mentioning that the result are specific to the code and even the computer running the benchmark and thus not useful for comparisons with other projects.

### Kimchi memory

> Similar to time, but with memory, the constraints are different in some ways, noise is not generally a problem and any computer cab be used, and also requires less sampling and should run faster. This is to some degree already implemented through a custom measurement for criterion, and in practice can be slower than measuring time, the instrumentation makes it run slower and the some limitations in criterion require to run the code more than necessary. With that in mind the best would be to run just a subset of these benchmark for regressions and let the others be run at discretion.

Reviewer 2: ```suggestion
Similar to time, but with memory, the constraints are different in some ways, noise is not generally a problem and any computer can be used, and also requires less sampling and should run faster. This is to some degree already implemented through a custom measurement for criterion, and in practice can be slower than measuring time, the instrumentation makes it run slower and the some limitations in criterion require to run the code more than necessary. With that in mind, the best would be to run just a subset of these benchmark for regressions and let the others be run at discretion.
```

Reviewer 2: Do you already have some ideas about which benchmarks would be marked as "regressions" or not?

# Benchmarking RFC

This document proposes a high level approach to benchmarking the crypto stack, mostly kimchi, snarky and pickles. The main concern is to set up an infrastructure where it is possible to integrate different kinds of benchmarks with minimal effort. While the specific benchmarks are not an immediate concern, a few benchmarks of interest will be mentioned later.

## Types of benchmarks

There are different metrics that we can collect, and different ways of presenting them depending of who is interested in them.

### Audience

We can differentiate to groups of people interested in benchmarks:

- Developers working in the crypto stack, mostly our crypto team, where the main interest is to avoid regressions or make performance improvements.

> - People working outside of it, mostly users of our libraries, but could also include decision makers comparing other options or our marketing team, here the interest is more about more general metrics that can be compared with the competition or used to learn how to most efficiently use our technology.

Reviewer 3: I don’t think this audience is entirely served by the current proposal. Decision makers for adopting this proof system stack at the application-level will want to understand the wall-clock time on commodity hardware of proving zkApps.

zkApp proofs (for now) are not proven in native execution even when they’re run server-side. Moreover, we probably need non-native benchmarks to help us internally decide how to prioritize building a zkapp flow that offloads some work to a native proving backend server-side.

Concretely, I think this translates to including in our benchmark matrix the following extra kinds of performance and memory tests:
{kimchi,pickles,snarky}-jsvm-{browser,server}

where “jsvm” is a mix of WASM and jsofocaml like SnarkyJS uses and browser means “run in the browser {V8 (Chrome), Safari, Firefox}” and server means running under a nodejs V8 environment. If we need to cut scope, just V8 for now in one of browser/server could be sufficient.

I am missing some context perhaps but I think for external audiences, these pickles web-based tests are either the most important benchmarks or second after native kimchi (for the purpose of being comparative against native proof systems).

Author: I added a bit at the end about leaving browser out of scope for now, but could be added at some point.

### Pickles memory

Similar to snarky, this will probably be mostly regression tests and maybe some example application to give users an idea of what to expect.

## Benchmarks

Here is a non exhaustive list of benchmarks.

### Kimchi hash chain

Benchmarking different aspects of a circuit proving a hash chain.

- The length of the hash chain can easily be a parameter allowing to measure different lengths. 
- We want memory and time for compilation, proving and verification, it will be implemented as several benchmarks but they will share most code.
- The number of constraints can be easily shown for each length.
- This benchmark with some specific length can be used as regression tests in CI.
- The full benchmark with several lengths can be shown at discretion of those interested running locally, or even added as optional to CI with some way to trigger it.

### Pickles hash chain

Same as kimchi hash chain but in pickles, we have less to cover here because we already have the kimchi ones, some time and memory benchmarks could be used for regressions and to observe the overhead, but having parameters like different circuits sizes may be redundant.

> A new benchmark possible would be to have an alternative circuit that uses recursion to make the hash chain instead of a single circuit, it could be interesting to have a comparison to the non recursive approach.

Reviewer 2: ```suggestion
A new  possible benchmark would be to have an alternative circuit that uses recursion to make the hash chain instead of a single circuit. It could be interesting to have a comparison to the non recursive approach.
```

### General kimchi benchmarks

> Some general benchmarks covering compilation, proving and verifying, both time and memory. Ideally all of them will parameterized with different circuit sizes and a subset can be used as regression tests. We could also add here benchmarks for the different operations of our polynomial commitments scheme.

Reviewer 2: ```suggestion
Some general benchmarks covering compilation, proving and verifying, both time and memory. Ideally all of them will be parameterized with different circuit sizes and a subset can be used as regression tests. We could also add here benchmarks for the different operations of our polynomial commitments scheme.
```

measuring memory but the most useful for us is peak memory, as this will tell
whether something can even be used in a particular environment. It can also help
finding errors related to unusual memory usage.  Memory usage is easier to
measure, as it does not appear to depend as much on the noise of the environment
as long as there is enough memory.

#### Hardware utilization

It can be of interest in some cases to measure how much of the available
resources are actually used, ideally the same computer will take half the time
with twice the resources, unfortunately this isn't always the case and sometimes
additional resources go unused.  This doesn't change that often, the main
utility of these benchmarks is just to detect low utilization so that it can be
improved.  Hardware utilization could include many things, GPU acceleration for
example if we were to start using it, but for now will only include CPU/core
utilization.

### Layers

There are 3 main layers of the stack that we want to consider, kimchi, snarky
and pickles. Given that they are implemented in different languages and exist in
different repositories special considerations each layer will be required.

> ## Where to run the benchmarks

Reviewer 3: Can we run the benchmarks within the profiling frontend that Viable Systems created? https://perf.ci.openmina.com/builds . I bet the velocity team can help support this. I’d love to see explanations on how that would work in this RFC and I’d recommend looking into before deciding on benchmarking libraries because I know they’ve already instrumented some of the rust and ocaml code. I’ll invite someone from viable systems to comment here to describe how it works from their side.

To start everyone with the resources should be able to run the benchmarks
locally, additionally we may want some of the benchmarks to run in CI to detect

> regressions. Some benchmarks may be considerably expensive to run, a decision

Author: I wasn't thinking of any in particular, it's more a consideration that we will have to do once in CI, relative to other benchmarks and our budget.
For each benchmark we will have to decide if it's worth having it in CI based on the running time we get.

will have to be made about which benchmarks to run and how often we should run
them.

> ## Parameters

Reviewer 3: Which explicitly will we parameterize and what will the parameters be?

Author: While I was thinking more about the general ability to have parameters, the only one I can think for now is circuit size.
It's basically running the same benchmark for different values of the parameter, for example running a prover time benchmark for circuits of size $2^n$ for $n$ in $[8..16]$.
I can make it more specific and list the parameters with only circuit size and then amend it if we want some other new parameters.

| kimchi instructions      | proof-systems | time     | no               | yes              | yes             | no               |
| kimchi memory            | proof-systems | memory   | yes              | yes              | yes             | no               |
| kimchi memory flamegraph | proof-systems | memory   | no               | yes              | no              | no               |
| snarky time              | mina/snarky*  | time     | yes              | yes              | yes             | yes              |
| snarky memory            | mina/snarky*  | memory   | yes              | yes              | yes             | no               |
| pickles time             | mina/pickles* | time     | yes              | yes              | yes             | yes              |
| pickles memory           | mina/pickles* | memory   | yes              | yes              | yes             | no               |

\* these have to be in Mina now because pickles is in Mina, and while snarky has
its own repository, using it requires a few things still in Mina, so that we
might want to consider some changes to make these benchmarks easier and probably
cheaper.

### Kimchi time

> One of the simplest measuring running time, can be a specific case or allow more

Reviewer 3: Which will we choose? A specific case? What is the case? Something more flexible? Can you describe it in detail?

Author: With specific case I mean without parameters. I think it would be a per benchmark decision, we could choose for example circuit size $2^{16}$ that is similar to the mina circuits or have more flexibility through parameters and benchmark a range of sizes.

flexibility by introducing parameters like for example different circuit sizes.

> We already have most of the infrastructure for this with

Reviewer 3: What is the remaining infrastructure beyond the criterion crate?

Author: In dependencies I don't we need anything we don't have already, there is the hardware, but I think that is not specific to this but more general to the whole RFC.
What remains is just some code, mostly thinking more in depth about the code structure now that there will be more than a single benchmark, creating utilities for that could help reduce code duplication, adding documentation and refactoring where necessary. Mostly small tasks that can be done in a day.

[criterion](https://crates.io/crates/criterion). We would ideally have a
dedicated computer. However, considering that there aren't that many PRs in
proof-systems, the requirements shouldn't be that high.

### Kimchi time flamegraph

This can be considered a variant of the [previous](#kimchi-time). Some more

> research is required but this information ideally can be generated from the

Reviewer 3: I think the time for the research is in this document. If it’s too complex to include in the scope of this RFC, then we can remove it for now and address it in another rfc

Author: Looked into it, at least for time flamegraphs it shouldn't be too hard, there is [cargo flamegraph](https://github.com/flamegraph-rs/flamegraph) that just wraps the binary and creates the flamegraph, it shouldn't require any change to the code.

already existing benchmarks with little to no changes.  This may not be as
interesting to have in CI, but can be useful to run locally for those working in
optimizations.

### Kimchi instructions

This is basically counting the CPU instructions that takes to run some code,
could be use as an alternative for CI given that it won't be affected by noise
and can work in any computer. It would also be faster by running the code only

> once. The implementation will require a bit more research but one option is the

Reviewer 3: Same with this one re research

Author: I think we can settle with `iai`, we already have it as dependency, it could require some more work that just running the time benchmark but should be a mere implementation detail.

crate [iai](https://crates.io/crates/iai), with some thinking it could be
possible to have this and the time benchmarks share the same code.  Is worth
mentioning that the result are specific to the code and even the computer
running the benchmark and thus not useful for comparisons with other projects.

### Kimchi memory

Similar to time, but with memory, the constraints are different in some ways,
noise is not generally a problem and any computer can be used, and also requires
less sampling and should run faster. This is to some degree already implemented
through a custom measurement for criterion, and in practice can be slower than
measuring time, the instrumentation makes it run slower and the some limitations

> in criterion require to run the code more than necessary. With that in mind, the

Reviewer 3: Which are the subset? Can you be specific?

Author: I don't think it can be specified at this point, it would depend on how much each benchmark costs, (both time and money), cost of all benchmarks, how much do we want to spend and what benchmarks seem more valuable relative to others.
We could start with running every single benchmark and removing the least valuables until we are under budget.

best would be to run just a subset of these benchmark for regressions and let
the others be run at discretion.

### Kimchi memory flamegraph

Similar to the time flamegraph, ideally can be made from the memory benchmark

> without much additional work.

Reviewer 3: What specifically needs to happen?

### Snarky time

This is similar to kimchi time, but the fact that it will be in the mina repo
should be considered, CI will run much more often and costs may be higher, this

> would be a reason to prefer to have benchmarks in kimchi when possible.

Reviewer 3: Do you recommend removing this then? Or if it’s part of kimchi, how?

Author: I think is just a reason to prefer having benchmarks in `proof-systems` when possible, but some benchmarks will have to be at this level for their very nature.

### Snarky memory

Here the main interest may be regressions, everything else can probably done at

> the kimchi level more precisely.

Reviewer 3: Agreed, what are the steps to achieve this?

Author: Not sure I understand the question, I decided to use `memtrace` since then for ocaml memory benchmarks, does that answer your question?

### Pickles time

This is similar to snarky time, mostly differing in the overhead of the
recursion. At this level beyond some regression we will probably have mostly

> application circuits being benchmarked.

Reviewer 3: Which applications and how will they be benchmarked specifically?

Author: I don't this section should that specific, for tools I decided to use `Core_bench` for ocaml time benchmarks, I can add that. About specific applications that would be in `## benchmarks`. This section is more about a common infrastructure/approach that can be reused for many benchmarks.

### Pickles memory

> Similar to snarky, this will probably be mostly regression tests and maybe some

Reviewer 3: What will they be exactly?

example application to give users an idea of what to expect.

## Benchmarks

Here is a non exhaustive list of benchmarks.

### Kimchi hash chain

> 

Reviewer 3: This section is a great start and has more detail, but there are still questions on precisely what we will do here? What will be used to determine the lengths that we set for the different runtimes? How will the tests be written? Where is the code, which libraries, etc

Author: The lengths chosen for regression tests are mostly arbitrary, the only constrains is CI cost, the process I can think of is choosing an arbitrary number and going up or down depending on how long it takes to run.
The code code should be Rust in the `proof-systems` repo, if that is precise enough, it shouldn't require additional dependencies over the already mentioned `criterion`.

Benchmarking different aspects of a circuit proving a hash chain.

- The length of the hash chain can easily be a parameter allowing to measure
different lengths.
- We want memory and time for compilation, proving and verification, it will be
implemented as several benchmarks but they will share most code.
- The number of constraints can be easily shown for each length.  
- This benchmark with some specific length can be used as regression tests in CI.
- The full benchmark with several lengths can be shown at discretion of those
interested running locally, or even added as optional to CI with some way to
trigger it.

### Pickles hash chain

Same as kimchi hash chain but in pickles, we have less to cover here because we
already have the kimchi ones, some time and memory benchmarks could be used for
regressions and to observe the overhead, but having parameters like different
circuits sizes may be redundant.  A new  possible benchmark would be to have an
alternative circuit that uses recursion to make the hash chain instead of a

> single circuit. It could be interesting to have a comparison to the non

Reviewer 3: Can you add specifics here too?

recursive approach.

> ### General kimchi benchmarks

Reviewer 3: Same here

Some general benchmarks covering compilation, proving and verifying, both time
and memory. Ideally all of them will be parameterized with different circuit
sizes and a subset can be used as regression tests. We could also add here
benchmarks for the different operations of our polynomial commitments scheme.

# Drawbacks

Beyond the possibility of spending the time in any other project the main
drawback I would be the cost, both in CI running time and dollars. This will
probably be a tradeoff between how much to spend and how much to run in CI.

# Rationale and alternatives

The alternative would be to handle benchmarks in a case by case basis, it may be
cheaper at short term to not have to design and implement this shared
infrastructure. Ultimately would be a matter of how much work will be saved
working on benchmarks. The suggested design may also incentivize the creation of
more benchmarks by lowering to barriers to create them.

# Prior art

Right now there are just a few benchmarks covering specific cases that can be
run for command line, mostly in the proof-systems repo. Almost all benchmarks
are also limited to measuring only time.

# Unresolved questions

> - While the tools for rust are mostly covered, specific tools for ocaml have yet

Reviewer 3: Do you intend on sorting out the ocaml parts before starting the project? If not, let’s remove them all from the RFC and make a new RFC later with the ocaml parts.
